### ISSUE TITLE
Ruins tab directory on various operations

### ISSUE THREAD

**Issue by levouh** [kyazdani42/nvim-tree.lua#239]
Not entirely sure what all I should include here, but I can reproduce this consistently with `nvim-tree.lua` installed, and it goes away as soon as I uninstall it, no other changes.

In order to reproduce:

1. Start `nvim`, note the directory of the current tab with `:echo getcwd(-1, tabpagenr())`.
2. Open a new tab, and `tcd` to a different directory.
3. Open a file in the second tab, and note that the `getcwd(-1, tabpagenr())` for the first tab also changes.

This also happens when you do things like `:G log -5 %` or whatever from `vim-fugitive`, essentially anything that triggers a `BufEnter` autocommand. I can see this by doing:

```
:9verbose e test/helpers.lua
```

in the second tab, which shows:

```
...
Executing BufEnter Autocommands for "*"
autocommand lua require'nvim-tree'.buf_enter()
...
```

If we look at this function, we can do something like:

```

function M.on_enter()
  local bufnr = api.nvim_get_current_buf()
  local bufname = api.nvim_buf_get_name(bufnr)
  local buftype = api.nvim_buf_get_option(bufnr, 'filetype')
  local ft_ignore = vim.g.nvim_tree_auto_ignore_ft or {}

  local stats = luv.fs_stat(bufname)
  local is_dir = stats and stats.type == 'directory'

  local disable_netrw = vim.g.nvim_tree_disable_netrw or 1
  local hijack_netrw = vim.g.nvim_tree_hijack_netrw or 1
  if is_dir then
    api.nvim_command('cd '..bufname)
  end
  local should_open = vim.g.nvim_tree_auto_open == 1
    and ((is_dir and (hijack_netrw == 1 or disable_netrw == 1)) or bufname == '')
    and not vim.tbl_contains(ft_ignore, buftype)

  print('bufnr=' .. vim.inspect(bufnr))
  print('bufname=' .. vim.inspect(bufname))
  print('stats=' .. vim.inspect(stats))
  print('is_dir=' .. vim.inspect(is_dir))
  print('should_open=' .. vim.inspect(should_open))

  lib.init(should_open, should_open)
end
```

which prints:

```
bufnr=1
bufname=""
stats=nil
is_dir=nil
should_open=false
```

noting that `bufnr=1` is actually on the first tab. Regardless, it doesn't seem to be a problem, as the call to `api.nvim_command`, various things in `lib.init(...)`, etc. all get skipped.

For reference, my configuration looks like:

```
vim.g.nvim_tree_ignore = {
    '.git',
    'node_modules',
    '*.pyc',
    '__pycache__',
}

vim.g.nvim_tree_auto_close = 1
vim.g.nvim_tree_indent_markers = 1
vim.g.nvim_tree_git_hl = 0

vim.g.nvim_tree_show_icons = {
    git = 0,
    folders = 1,
    files = 1,
}

vim.g.nvim_tree_bindings = {
    ['-'] = get_func('close_node'),
    ['<BS>'] = get_func('dir_up'),
    ['<CR>'] = get_func('edit'),
    ['dd'] = get_func('remove'),
    ['o'] = get_func('create'),
    ['p'] = get_func('paste'),
    ['yy'] = get_func('copy'),
}

vim.g.nvim_tree_icons = {
    default = '',
    symlink = '',
    git = {
        unstaged = '✗',
        staged = '✓',
        unmerged = '',
        renamed = '➜',
        untracked = '★'
    },
    folder = {
        default = '',
        open = '',
        empty = '',
        empty_open = '',
        symlink = '',
    }
}
```

---

So my question then is, what exactly is happening here? And what else can I do to help debug this? I love this plugin, but with my usage of tabs, this is a deal-breaker if my directories are going to be messed up.

**Comment by kyazdani42** [comment]
Hi and thanks for this detailed report :)
I believe path handling is very clunky, mostly because it uses the main vim root path.
You could check #234, i believe it is related somehow.
I'll do a PR soon to fix these paths issues, these seems to get in the way of many people.

**Comment by levouh** [comment]
I will certainly check that out. Also, this plugin is amazing and I appreciate all the effort that has been put into it.

I've been messing with it on my end when I have time, so I'll update if I find anything useful. Since this is on your radar, should I keep this issue open or do you have another one to track this effort?

**Comment by kyazdani42** [comment]
Leave it open i'm currently working on a small refactoring that will probably fix this :)

**Comment by levouh** [comment]
Not entirely sure what your plans are @kyazdani42, but changing the two calls to `nvim_command` that issue a `cd` to either `tcd` or `lcd` fixes this problem. I've used a local fork for a while now with this change and everything seems to work with it.

This might not align well with the refactor you've been talking about, nor the overall usage of the directory-changes in the first place, or specific use cases, etc.

It has worked for me, so I'd be curious on your feedback as it is a _really_ simple change to make.

**Comment by kyazdani42** [comment]
indeed we should use `lcd` and not `cd` to avoid altering the neovim root. Also this might help with the `DirChanged` autocmd ? I'll make a small PR

**Comment by levouh** [comment]
Note @kyazdani42 that this isn't actually fixed with #313. That might handle the view of the tree on a tab page correctly, but the issue mentioned above still exists. More specifically, the working directory of a tab is not preserved when dealing with multiple of them as this plugin tends to `cd` when it should probably `tcd` or `lcd` instead.

I'm not entirely sure why the changing of directories is necessary in the first place, or the use cases, etc. so what I've mentioned above obviously might not be as simple as I've made it out to be.

**Comment by kyazdani42** [comment]
woups didn't read the discussion, i think it really depends on user preference actually. Thing is, i do not handle multiple workspaces so using tcd might result in undefined behavior as the buffer is shared between windows (and thus between tabpages).
I'm not yet sure about the behavior i want to implement for path handling. Basically, tcl and lcd are not moving nvim base cwd which i personnally want (because of lsp's and such), but this has been discussed in #234. Maybe just continue discussion there, the title wasn't adapted anymore anyway :)
